tests/vklayertests_imageless_framebuffer.cpp

```
@@ -1664,6 +1664,7 @@ TEST_F(VkLayerTest, DescriptorUpdateTemplateEntryWithInlineUniformBlock) {
         m_device_extension_names.push_back(VK_KHR_PUSH_DESCRIPTOR_EXTENSION_NAME);
     } else {
         printf("%s Push Descriptor Extension not supported, push descriptor cases skipped.\n", kSkipPrefix);
+        return;
```

Actually, I'm not sure why lines 1663-1668 are here. Looks like Push descriptors aren't used in the test at all. @ziga-lunarg?

---

Looks like a copy-paste error and should be removed

---

Thanks @ziga-lunarg. @nisse-work, would you mind removing 1659-1668 as part of this change?

---

I gave it a try, but then the test crash (still on intel mesa). Complete backtrace below. Interesting points:

Crash is in the test's call to InitState. A bit down, it gets to ValidationStateTracker::PostCallRecordCreateDevice.  This calls

    GetPhysicalDeviceExtProperties(gpu, dev_ext.vk_ext_inline_uniform_block, &phys_dev_props->inline_uniform_block_props);

(this is the third call in a group of four calls to GetPhysicalDeviceExtProperties). This gets to code in the GetPhysicalDeviceExtProperties template in state_tracker.h;

            if (api_version < VK_API_VERSION_1_1) {
                auto prop2 = LvlInitStruct<VkPhysicalDeviceProperties2>(ext_prop);
                DispatchGetPhysicalDeviceProperties2KHR(gpu, &prop2);
            }

And then DispatchGetPhysicalDeviceProperties2KHR calls

    layer_data->instance_dispatch_table.GetPhysicalDeviceProperties2KHR(physicalDevice, pProperties);

which jumps to a null pointer. To me, this is puzzling. I see that the test includes an explicit check for VK_EXT_INLINE_UNIFORM_BLOCK_EXTENSION_NAME, and it thinks this extension is available. The vulkaninfo tool agrees, and lists

        VK_EXT_inline_uniform_block               : extension revision 1

``` 
#0  0x0000000000000000 in ?? ()
#1  0x00007fffeaf31805 in DispatchGetPhysicalDeviceProperties2KHR (
    physicalDevice=0x55555a141360, pProperties=0x7fffffffb820)
    at /home/nisse/hack/Vulkan-ValidationLayers/layers/generated/layer_chassis_dispatch.cpp:4675
#2  0x00007fffeb02ece6 in ValidationStateTracker::GetPhysicalDeviceExtProperties<VkPhysicalDeviceInlineUniformBlockPropertiesEXT> (this=0x55555a05c9a0, gpu=0x55555a141360, 
    enabled=kEnabledByCreateinfo, ext_prop=0x55555a2a9878)
    at /home/nisse/hack/Vulkan-ValidationLayers/layers/state_tracker.h:1338
#3  0x00007fffeb001a22 in ValidationStateTracker::PostCallRecordCreateDevice (
    this=0x55555a05c9a0, gpu=0x55555a141360, pCreateInfo=0x7fffffffce30, 
    pAllocator=0x0, pDevice=0x7fffffffcde8, result=VK_SUCCESS)
    at /home/nisse/hack/Vulkan-ValidationLayers/layers/state_tracker.cpp:1374
#4  0x00007fffeb0fec42 in CoreChecks::PostCallRecordCreateDevice (this=0x55555a05c9a0, 
    gpu=0x55555a141360, pCreateInfo=0x7fffffffce30, pAllocator=0x0, 
    pDevice=0x7fffffffcde8, result=VK_SUCCESS)
    at /home/nisse/hack/Vulkan-ValidationLayers/layers/core_validation.cpp:2680
#5  0x00007fffeae2c552 in vulkan_layer_chassis::CreateDevice (gpu=0x55555a141360, 
    pCreateInfo=0x7fffffffce30, pAllocator=0x0, pDevice=0x7fffffffcde8)
    at /home/nisse/hack/Vulkan-ValidationLayers/layers/generated/chassis.cpp:515
#6  0x00007ffff7815390 in ?? () from /lib/x86_64-linux-gnu/libvulkan.so
#7  0x00007ffff7815fba in ?? () from /lib/x86_64-linux-gnu/libvulkan.so
#8  0x00007ffff7818f0f in vkCreateDevice () from /lib/x86_64-linux-gnu/libvulkan.so
#9  0x00005555568ae2fa in vk_testing::Device::init (this=0x55555a05fea0, info=...) at /home/nisse/hack/Vulkan-ValidationLayers/tests/vktestbinding.cpp:294
#10 0x00005555568ae239 in vk_testing::Device::init (this=0x55555a05fea0, extensions=std::vector of length 2, capacity 2 = {...}, features=0x0, create_device_pnext=0x0) at /home/nisse/hack/Vulkan-ValidationLayers/tests/vktestbinding.cpp:288
#11 0x0000555556894f13 in VkDeviceObj::VkDeviceObj (this=0x55555a05fea0, id=0, obj=0x55555a062e00, extension_names=std::vector of length 2, capacity 2 = {...}, features=0x0, create_device_pnext=0x0) at /home/nisse/hack/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:1032
#12 0x00005555568934c8 in VkRenderFramework::InitState (this=0x5555599b5360, features=0x0, create_device_pnext=0x0, flags=2) at /home/nisse/hack/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:596
#13 0x000055555678d8ca in VkLayerTest_DescriptorUpdateTemplateEntryWithInlineUniformBlock_Test::TestBody (this=0x5555599b5360) at /home/nisse/hack/Vulkan-ValidationLayers/tests/vklayertests_imageless_framebuffer.cpp:1671
#14 0x00007ffff7e05046 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x5555599b5360, method=&virtual testing::Test::TestBody(), location=0x7ffff7e16e0b "the test body") at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2443
#15 0x00007ffff7dfef9b in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=0x5555599b5360, method=&virtual testing::Test::TestBody(), location=0x7ffff7e16e0b "the test body") at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2479
#16 0x00007ffff7ddef64 in testing::Test::Run (this=0x5555599b5360) at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2517
#17 0x00007ffff7ddf803 in testing::TestInfo::Run (this=0x55555997e890) at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2693
#18 0x00007ffff7ddfe53 in testing::TestCase::Run (this=0x55555991df30) at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2811
#19 0x00007ffff7dea75c in testing::internal::UnitTestImpl::RunAllTests (this=0x55555991dbe0) at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:5177
#20 0x00007ffff7e0619f in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x55555991dbe0, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x7ffff7dea4b4 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x7ffff7e177e8 "auxiliary test code (environments or event listeners)") at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2443
#21 0x00007ffff7dffec5 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x55555991dbe0, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x7ffff7dea4b4 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x7ffff7e177e8 "auxiliary test code (environments or event listeners)") at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2479
#22 0x00007ffff7de9320 in testing::UnitTest::Run (this=0x7ffff7e3b860 <testing::UnitTest::GetInstance()::instance>) at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:4786
#23 0x00005555563cb9cd in RUN_ALL_TESTS () at /home/nisse/hack/Vulkan-ValidationLayers/external/googletest/googletest/include/gtest/gtest.h:2341
#24 0x00005555563c757b in main (argc=1, argv=0x7fffffffdf48) at /home/nisse/hack/Vulkan-ValidationLayers/tests/layer_validation_tests.cpp:3690
```

---

@jeremyg-lunarg @mikes-lunarg Any chance this needs a similar fix as #3156 at https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/0c09bc913497708c59ecb1349d6b83b02be255fd/layers/state_tracker.h#L1336?

i.e., I'm wondering if we can just change that condition to:
```cpp
if (device_extensions.vk_khr_get_physical_device_properties2) {
```
based on Mike's [previous comment](https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/3156#issuecomment-897193280)?

---

It seems that doesn't help. It seems *both* potential code paths (DispatchGetPhysicalDeviceProperties2KHR and DispatchGetPhysicalDeviceProperties2) jump via null pointers. If I examine the device_extensions, I see only three enabled entries,

    vk_ext_debug_utils = kEnabledByCreateinfo, 
    vk_ext_inline_uniform_block = kEnabledByCreateinfo
    vk_khr_descriptor_update_template = kEnabledByCreateinfo

And the caller (ValidationStateTracker::PostCallRecordCreateDevice) passes vk_ext_inline_uniform_block (true) as the enabled argument, which is why this call crashes while the previous two don't (they get a false value for enabled).

And api_version == VK_API_VERSION_1_0. 

So my understanding is there are two options to make this work:

1. Fix this code to skip calling any variant of VkGetPhysicalDeviceProperties2 when api version is < 1.1 and the extension isn't present. Not sure if there's a reasonable way to fail, could it work to just leave the corresponding properties struct default initialized? Most of the structs appear to be members of phys_dev_props, and I have no idea which code depends on these values. And I guess falling back to VkGetPhysicalDeviceProperties (no "2") can't work?

2. Use api version 1.1 or later for all tests that depend on this call (which would be all tests that enables any of the extensions with a corresponding enable flag passed to GetPhysicalDeviceExtProperties. But that won't help applications that use api version 1.0 together with validation layers, right?

---

>Any chance this needs a similar fix as #3156

No, this looks like something else.

I think the problem here is how the test is enabling VK_KHR_get_physical_device_properties2. If the test doesn't find `VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_SPEC_VERSION` version (2) of the extension it just doesn't enable it. This is wrong, that extension is required when enabling VK_EXT_inline_uniform_block. I bet @nisse-work has version 1 on their machine.

Questions:
- Why does the test need that revision of the VK_KHR_get_physical_device_properties2 extension?
- Why don't extension dependency validation errors get flagged by the test framework? There should have also been an error because VK_EXT_inline_uniform_block requires VK_KHR_maintenance1.
- Why is the layer chassis calling a null fp? I thought we checked for null pointers in the dispatch table and swapped them out for stubs.

---

For bullet 2, it _might_ be that in the LVTs, errors are ignored unless they're being looked for by the errorMonitor

---

> For bullet 2, it _might_ be that in the LVTs, errors are ignored unless they're being looked for by the errorMonitor.

That makes sense. A few months ago we found an issue because that the test framework doesn't enforce a clean teardown. It seems that it doesn't check for a clean setup either.

---

My guess would be it doesn't. Trying this out locally, using version 1 seems to "work." However, if I put `m_errorMonitor->ExpectSuccess()` at the top of the test followed by `m_errorMonitor->VerifyNotFound()` right before the "expected failures," I can then see that CreateDevice does indeed fail due to `VK_KHR_maintenance1` missing. This seems like some breakage in the testing framework :grimacing:

---

Started on #3178 to address the test framework behavior. Unfortunately this is a lot slower going than I expected. Hopefully I'll have this done by tomorrow.